File: src/java/main/org/apache/zookeeper/server/quorum/ObserverZooKeeperServer.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.zookeeper.server.quorum;

import java.io.IOException;
import java.util.concurrent.ConcurrentLinkedQueue;

import org.apache.log4j.Logger;
import org.apache.zookeeper.server.FinalRequestProcessor;
import org.apache.zookeeper.server.Request;
import org.apache.zookeeper.server.RequestProcessor;
import org.apache.zookeeper.server.SyncRequestProcessor;
import org.apache.zookeeper.server.ZKDatabase;
import org.apache.zookeeper.server.persistence.FileTxnSnapLog;

/**
 * A ZooKeeperServer for the Observer node type. Not much is different, but
 * we anticipate specializing the request processors in the future. 
 *
 */
public class ObserverZooKeeperServer extends LearnerZooKeeperServer {
    private static final Logger LOG =
        Logger.getLogger(ObserverZooKeeperServer.class);        
    
    /*
     * Request processors
     */
    private CommitProcessor commitProcessor;
    private SyncRequestProcessor syncProcessor;
    
    /*
     * Pending sync requests
     */
    ConcurrentLinkedQueue<Request> pendingSyncs = 
        new ConcurrentLinkedQueue<Request>();
        
    ObserverZooKeeperServer(FileTxnSnapLog logFactory, QuorumPeer self,
            DataTreeBuilder treeBuilder, ZKDatabase zkDb) throws IOException {
        super(logFactory, self.tickTime, self.minSessionTimeout,
                self.maxSessionTimeout, treeBuilder, zkDb, self);
    }
    
    public Observer getObserver() {
        return self.observer;
    }
    
    @Override
    public Learner getLearner() {
        return self.observer;
    }       
    
    /**
     * Unlike a Follower, which sees a full request only during the PROPOSAL
     * phase, Observers get all the data required with the INFORM packet. 
     * This method commits a request that has been unpacked by from an INFORM
     * received from the Leader. 
     *      
     * @param request
     */
    public void commitRequest(Request request) {     
        commitProcessor.commit(request);        
    }
    
    /**
     * Set up the request processors for an Observer:
     * firstProcesor->commitProcessor->finalProcessor
     */
    @Override
    protected void setupRequestProcessors() {      
        // We might consider changing the processor behaviour of 
        // Observers to, for example, remove the disk sync requirements.
        // Currently, they behave almost exactly the same as followers.
        RequestProcessor finalProcessor = new FinalRequestProcessor(this);
        commitProcessor = new CommitProcessor(finalProcessor,
                Long.toString(getServerId()), true);
        commitProcessor.start();
        firstProcessor = new ObserverRequestProcessor(this, commitProcessor);
        ((ObserverRequestProcessor) firstProcessor).start();
        syncProcessor = new SyncRequestProcessor(this,
                new SendAckRequestProcessor(getObserver()));
        syncProcessor.start();
    }
    
    /*
     * Process a sync request
     */
    synchronized public void sync(){
        if(pendingSyncs.size() ==0){
            LOG.warn("Not expecting a sync.");
            return;
        }
                
        Request r = pendingSyncs.remove();
        commitProcessor.commit(r);
    }
    
    @Override
    public String getState() {
        return "observer";
    };    
}

File: zookeeper/server/quorum/ObserverZooKeeperServer.cpp
/*
 * ObserverZooKeeperServer.cpp
 *
 *  Created on: 2017-11-22
 *      Author: cxxjava@163.com
 */

#include "./Observer.hh"
#include "./QuorumPeer.hh"
#include "./ObserverZooKeeperServer.hh"
#include "./ObserverRequestProcessor.hh"
#include "../FinalRequestProcessor.hh"

namespace efc {
namespace ezk {

sp<ELogger> ObserverZooKeeperServer::LOG = ELoggerManager::getLogger("ObserverZooKeeperServer");

ObserverZooKeeperServer::ObserverZooKeeperServer(sp<FileTxnSnapLog> logFactory, QuorumPeer* self,
		sp<ZKDatabase> zkDb) THROWS(EIOException) :
	LearnerZooKeeperServer(logFactory, self->tickTime, self->minSessionTimeout,
			self->maxSessionTimeout, zkDb, self) {
	syncRequestProcessorEnabled = self->getSyncEnabled();
	LOG->info(EString("syncEnabled =") + syncRequestProcessorEnabled);
}

sp<Observer> ObserverZooKeeperServer::getObserver() {
	return self->observer;
}

void ObserverZooKeeperServer::setupRequestProcessors() {
	// We might consider changing the processor behaviour of
	// Observers to, for example, remove the disk sync requirements.
	// Currently, they behave almost exactly the same as followers.
	sp<RequestProcessor> finalProcessor = new FinalRequestProcessor(shared_from_this());
	commitProcessor = new CommitProcessor(finalProcessor,
			ELLong::toString(getServerId()), true,
			getZooKeeperServerListener());
	EThread::setDaemon(commitProcessor, true); //!!!
	commitProcessor->start();

	sp<ObserverRequestProcessor> orp = new ObserverRequestProcessor(dynamic_pointer_cast<ObserverZooKeeperServer>(shared_from_this()), commitProcessor);
	firstProcessor = orp;
	EThread::setDaemon(orp, true); //!!!
	orp->start();
	/*
	 * Observer should write to disk, so that the it won't request
	 * too old txn from the leader which may lead to getting an entire
	 * snapshot.
	 *
	 * However, this may degrade performance as it has to write to disk
	 * and do periodic snapshot which may double the memory requirements
	 */
	if (syncRequestProcessorEnabled) {
		syncProcessor = new SyncRequestProcessor(shared_from_this(), null);
		EThread::setDaemon(syncProcessor, true); //!!!
		syncProcessor->start();
	}
}

sp<Learner> ObserverZooKeeperServer::getLearner() {
	return self->observer;
}

void ObserverZooKeeperServer::commitRequest(sp<Request> request) {
	if (syncRequestProcessorEnabled) {
		// Write to txnlog and take periodic snapshot
		syncProcessor->processRequest(request);
	}
	commitProcessor->commit(request);
}

void ObserverZooKeeperServer::sync(){
	SYNCHRONIZED(this) {
		if(pendingSyncs.size() ==0){
			LOG->warn("Not expecting a sync.");
			return;
		}

		sp<Request> r = pendingSyncs.remove();
		commitProcessor->commit(r);
	}}
}

void ObserverZooKeeperServer::shutdown() {
	SYNCHRONIZED(this) {
		if (!canShutdown()) {
			LOG->debug("ObserverZooKeeper server is not running, so not proceeding to shutdown!");
			return;
		}
		LearnerZooKeeperServer::shutdown();
		if (syncRequestProcessorEnabled && syncProcessor != null) {
			syncProcessor->shutdown();
		}
	}}
}

} /* namespace ezk */
} /* namespace efc */
